src/Mod/Sketcher/Gui/CommandSketcherTools.cpp

```diff
@@ -983,50 +983,61 @@ void CmdSketcherRestoreInternalAlignmentGeometry::activated(int iMsg)
-            (it->size() > 12 && it->substr(0,12) == "ExternalEdge")) {
-            int GeoId;
-            if (it->substr(0,4) == "Edge")
-               GeoId = std::atoi(it->substr(4,4000).c_str()) - 1;
-            else
-               GeoId = -std::atoi(it->substr(12,4000).c_str()) - 2;
-
-            const Part::Geometry *geo = Obj->getGeometry(GeoId);
-            // Only for supported types
-            if (geo->getTypeId() == Part::GeomEllipse::getClassTypeId() ||
-                geo->getTypeId() == Part::GeomArcOfEllipse::getClassTypeId() ||
-                geo->getTypeId() == Part::GeomArcOfHyperbola::getClassTypeId() ||
-                geo->getTypeId() == Part::GeomArcOfParabola::getClassTypeId() ||
-                geo->getTypeId() == Part::GeomBSplineCurve::getClassTypeId()) {
+    std::vector<int> SubGeoIds(SubNames.size());
+    std::transform(SubNames.begin(), SubNames.end(), SubGeoIds.begin(),
+                   [Obj](const std::string& SubName) {
+                       int GeoId;
+                       Sketcher::PointPos PosId;
+                       getIdsFromName(SubName, Obj, GeoId, PosId);
+                       return GeoId;
+                   });
+
+    SubGeoIds.erase(std::unique(SubGeoIds.begin(), SubGeoIds.end()), SubGeoIds.end());
+
+    // Only for supported types
+    SubGeoIds.erase(
+        std::remove_if(SubGeoIds.begin(), SubGeoIds.end(),
+                       [Obj](const auto& GeoId) {
+                           const Part::Geometry *geo = Obj->getGeometry(GeoId);
+                           bool hasInternalGeo = geo &&
+                               (geo->getTypeId() == Part::GeomEllipse::getClassTypeId() ||
+                                geo->getTypeId() == Part::GeomArcOfEllipse::getClassTypeId() ||
+                                geo->getTypeId() == Part::GeomArcOfHyperbola::getClassTypeId() ||
+                                geo->getTypeId() == Part::GeomArcOfParabola::getClassTypeId() ||
+                                geo->getTypeId() == Part::GeomBSplineCurve::getClassTypeId());
+                           return !hasInternalGeo; // so it's removed
+                       }),
+        SubGeoIds.end());
+
+    // Handle highest GeoIds first to minimize GeoIds changing
+    // TODO: this might not completely resolve GeoIds changing completely
+    std::sort(SubGeoIds.begin(), SubGeoIds.end(), std::greater<int>());
 
-                int currentgeoid = Obj->getHighestCurveIndex();
+    // go through the selected subelements
+    for (const auto& GeoId : SubGeoIds) {
+        int currentgeoid = Obj->getHighestCurveIndex();
 
-                try {
-                    Gui::Command::openCommand(QT_TRANSLATE_NOOP("Command", "Exposing Internal Geometry"));
-                    Gui::cmdAppObjectArgs(Obj, "exposeInternalGeometry(%d)", GeoId);
+        try {
+            Gui::Command::openCommand(QT_TRANSLATE_NOOP("Command", "Exposing Internal Geometry"));
```

Comment:
Having this here means you will create a transaction for each geometry, which will makes undoing horrible. :) Should be done before for loop.

Reply:
This behavior is the same as before, though. Are you sure we want to change this?

Reply:
Actually I didn't check. :) Just make a "reading" code review.
If I'm correct (and actually a transaction is created for each geometry), I think it makes no sense. Only one transaction shall be stored. ;)

Reply:
https://github.com/FreeCAD/FreeCAD/blob/d1fc159f4c05feb455b324b4e32821817c30852a/src/Mod/Sketcher/Gui/CommandSketcherTools.cpp#L1008

Reply:
@0penBrain you were correct that only one transaction is stored (though why I do not understand).